packages/core/echo/echo-schema/src/query/query.ts

```diff
@@ -2,13 +2,91 @@
 // Copyright 2024 DXOS.org
 //
 
+import { SpaceId } from '@dxos/keys';
 import { Schema as S } from '@effect/schema';
 
+//
+// Inspired by filter.proto.
+//
+
+// TODO(dmaretskyi): Move out.
+const EchoIdSchema = S.String;
+
+// TODO(dmaretskyi): Move out.
+const SpaceIdSchema = S.String as any as S.Schema<SpaceId>;
+
+const DxnSchema = S.String.pipe(S.pattern(/^dxn:/));
+
+// TODO(dmaretskyi): Consider moving this to echo protocol.
+
 /**
  * ECHO query object.
  */
+const FilterSchema = S.Struct({
+  /**
+   * Matches specific object id.
+   */
+  objectId: S.optional(S.Array(EchoIdSchema)),
+
+  /**
+   * Matches object type.
+   */
+  type: S.optional(S.Array(DxnSchema)),
+
+  // TODO(dmaretskyi): How do we handle matching nested properties?
+  /*
+  For example if we want to match `address.city`,
+  would it be: `{ address: { city: 'Bangkok' } }` or `{ 'address.city': 'Bangkok' }`?
+
+  The second one is ambiguous because it could be a nested object or a string key.
+  The first one is not clear on whether it allows extra properties in the `address` object.
+
+  */
+  properties: S.optional(S.Record({ key: S.String, value: S.Any })),
+
+  /**
+   * Full text search.
+   */
+  text: S.optional(S.String),
+
+  /**
+   * This filter should be negated.
+   */
+  negated: S.optional(S.Boolean),
+
+  /**
+   * All additional filters that should be satisfied for this filter to pass.
+   * Additionally any criteria defined in this filter (e.g. `type`, `properties`) are also required to be satisfied.
+   */
+  and: S.optional(S.Array(S.suspend((): S.Schema<FilterType> => FilterSchema))),
+
+  /**
+   * Any of the additional filters should be satisfied for this filter to pass.
+   * Additionally any criteria defined in this filter (e.g. `type`, `properties`) are also required to be satisfied.
+   */
+  or: S.optional(S.Array(S.suspend((): S.Schema<FilterType> => FilterSchema))),
+});
+
+interface FilterType extends S.Schema.Type<typeof FilterSchema> {}
+
 const QuerySchema = S.Struct({
-  type: S.String,
+  /**
+   * Query filter.
+   */
+  filter: FilterSchema,
+
+  /**
+   * Space to query.
+   */
+  spaces: S.optional(S.Array(SpaceIdSchema)),
+
+  /**
+   * Eagerly load data from objects referenced by the references set in those specific properties.
+   *
+   * @example `{ org: true }` - loads data from the reference in the `org` property.
+   * @example `{ org: { manager: true } }` - recursively loads data from `org` reference and then from the `manager` reference.
+   */
+  include: S.optional(S.Record({ key: S.String, value: S.Any })),
```

Comment:
specify response shape?: https://graphql.org/learn/response


`
type: 'dxn:type:example.com/type/contact',
// Get nested orgs and projects.
response: {
  org: {
    projects: {}
  }
}
`

or  
- https://neo4j.com/docs/java-reference/current/traversal-framework/
- https://neo4j.com/docs/cypher-manual/current/subqueries/existential

```
MATCH (person:Person)
WHERE EXISTS {
  MATCH (person)-[:HAS_DOG]->(dog:Dog)
  WHERE EXISTS {
    MATCH (dog)-[:HAS_TOY]->(toy:Toy)
    WHERE toy.name = 'Banana'
  }
}
```

perhaps the intuition is to start with the DSL test?